frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx

```diff
@@ -16,150 +16,724 @@
 
-  const generateSpec = useCallback((): any => {
-    const spec = JSON.parse(
-      replaceTemporaryColors(figure.spec, theme, element.theme)
-    )
-    const initialHeight = spec.layout.height
-    const initialWidth = spec.layout.width
-
-    if (isFullScreen) {
-      spec.layout.width = width
-      spec.layout.height = height
-    } else if (element.useContainerWidth) {
-      spec.layout.width = width
-    } else {
-      spec.layout.width = initialWidth
-      spec.layout.height = initialHeight
+  // Load the initial figure spec from the element message
+  const initialFigureSpec = useMemo<PlotlyFigureType>(() => {
+    if (!element.spec) {
+      return {
+        layout: {},
+        data: [],
+        frames: undefined,
+      }
     }
-    if (element.theme === "streamlit") {
-      applyStreamlitTheme(spec, theme)
-    } else {
-      // Apply minor theming improvements to work better with Streamlit
-      spec.layout = layoutWithThemeDefaults(spec.layout, theme)
+
+    return JSON.parse(element.spec)
+    // We want to reload the initialFigureSpec object whenever the element id changes
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [element.id, element.spec])
+
+  const [plotlyFigure, setPlotlyFigure] = useState<PlotlyFigureType>(() => {
+    // If there was already a state with a figure using the same id,
+    // use that to recover the state. This happens in some situations
+    // where a component un-mounts and mounts again.
+    const initialFigureState = widgetMgr.getElementState(element.id, "figure")
+    if (initialFigureState) {
+      return initialFigureState
+    }
+    return applyTheming(initialFigureSpec, element.theme, theme)
+  })
+
+  const isSelectionActivated = element.selectionMode.length > 0 && !disabled
+  const isLassoSelectionActivated = element.selectionMode.includes(
+    PlotlyChartProto.SelectionMode.LASSO
+  )
+  const isBoxSelectionActivated = element.selectionMode.includes(
+    PlotlyChartProto.SelectionMode.BOX
+  )
+  const isPointsSelectionActivated = element.selectionMode.includes(
+    PlotlyChartProto.SelectionMode.POINTS
+  )
+
+  const plotlyConfig = useMemo(() => {
+    if (!element.config) {
+      // If there is no config, return an empty object
+      return {}
     }
 
-    return spec
+    const config = JSON.parse(element.config)
+
+    // Customize the plotly toolbar:
+    if (!disableFullscreenMode) {
+      // Add a fullscreen button to the plotly toolbar:
+      config.modeBarButtonsToAdd = [
+        {
+          name: isFullScreen ? "Close fullscreen" : "Fullscreen",
+          icon: isFullScreen
+            ? FULLSCREEN_COLLAPSE_ICON
+            : FULLSCREEN_EXPAND_ICON,
+          click: () => {
+            if (isFullScreen && collapse) {
+              collapse()
+            } else if (expand) {
+              expand()
+            }
+          },
+        },
+      ]
+    }
+
+    if (!config.modeBarButtonsToRemove && isSelectionActivated) {
```

Comment:
nit: we do want to apply these toolbar modifications in all cases (basically implementing what is suggested in this [project](https://www.notion.so/snowflake-corp/Rethink-Plotly-toolbar-f0b19b1e31494d5a9e3d224e2b93e301)). So its correct to apply this always:

```suggestion
    if (!config.modeBarButtonsToRemove) {
```

Reply:
also see the part below:

```python
if (!isSelectionActivated) {
        // Remove lasso & select buttons in read-only charts:
        modeBarButtonsToRemove.push("lasso2d", "select2d")
 }
```